drgn_tools/kernfs_memcg.py

```diff
@@ -0,0 +1,224 @@
+    return mem_cgroup_subsys.root.cgrp.nr_descendants.value_() + 1
+
+
+def get_num_dying_mem_cgroups(prog: Program) -> int:
+    """
+    Get number of inactive or dying mem cgroups.
+    """
+    mem_cgroup_subsys = prog["cgroup_subsys"][_MEMORY_CGRP_ID]
+    return mem_cgroup_subsys.root.cgrp.nr_dying_descendants.value_()
+
+
+# By default we scan all pages, that have memcg ref
+# but if max_pages is specified then we bail out
+# after getting those many pages or scanning all
+# pages , whichever happens first
+def dump_page_cache_pages_pinning_cgroups(
+    prog: Program, max_pages: int = 10000
+):
+    """
+    Dump page-cache pages that have reference to a mem-cgroup.
+    The ouput also contains information such as the cgroup that is pinned, its flags
+    (to indicate current state of cgroup) and file cached by this page.
+
+    :params: max_pages: specify how many pages to find. By default first 10000 such
+    pages are listed. Use 0 to list all such pages.
+
+    """
+    PG_slab_mask = 1 << prog.constant("PG_slab")
+    mem_cgroup_root = prog["cgroup_subsys"][_MEMORY_CGRP_ID].root
+    total_count = 0
+    found_count = 0
+    for page in for_each_page(prog):
+        total_count = total_count + 1
+        try:
+            # Ignore slab pages
+            if page.flags & PG_slab_mask:
+                continue
+            # Ignore non page-cache pages
+            if not page.mapping:
+                continue
+            try:
+                mem_cgroup = page.mem_cgroup
+            except AttributeError:
+                mem_cgroup = page.memcg_data
+
+            if not mem_cgroup.value_() or mem_cgroup.value_() & 3:
+                continue
+            cgroup_subsys_state = cast(
+                "struct cgroup_subsys_state *", mem_cgroup
+            )
+            if cgroup_subsys_state.cgroup.root == mem_cgroup_root:
+                found_count = found_count + 1
+                cgrp = cgroup_subsys_state.cgroup
+                address_space = page.mapping
+                inode = address_space.host
+                if inode_path(inode) is None:
+                    continue
+                dentry = container_of(
+                    inode.i_dentry.first, "struct dentry", "d_u.d_alias"
+                )
+                path = dentry_path_any_mount(dentry).decode()
+                cgroup_state = decode_css_flags(cgrp.self.address_of_())
+                print(
+                    f"page: 0x{page.value_():x} cgroup: {cgroup_path(cgrp).decode()} state: {cgroup_state} path: {path}\n"
+                )
+                if max_pages and found_count == max_pages:
+                    break
+        except FaultError:
+            continue
+
+    print(
+        f"Scanned {total_count} pages, found {found_count} pages with memory cgroup refs."
+    )
+
+
+class PagesPinningMemcgroups(CorelensModule):
+    """Print information related to pages, that are pinning memcgroup(s)"""
+
+    name = "kernfs_memcg"
+
```

Comment:
Is this the best name? Or something maybe more like "pages-pinning-memcg"?

Also, would it be worth adding:
```
run_when = "never"
```

Here, so that this only gets run if the user explicitly asks for this module? It seems to me that the output is less useful if you only scan a partial number of pages, and so I don't know that it would really be useful to have it run by default.

Reply:
> Is this the best name? Or something maybe more like "pages-pinning-memcg"?
> 
How about just memcg or memcgroups. pages-pinning-memcg is one specific debug scenario handled by this module currently but in future I intend to extend it for other memcgroup related debuggings as well
> Also, would it be worth adding:
> 
> ```
> run_when = "never"
> ```
> 
> Here, so that this only gets run if the user explicitly asks for this module? It seems to me that the output is less useful if you only scan a partial number of pages, and so I don't know that it would really be useful to have it run by default.
Makes sense. I will modify the module to just report number of zombie cgroups and if then user thinks that page scanning is needed, he/she can run it separately. Most of the vmcores will not have significant number of zombie memcgroups , so it does not seem like a good idea to do the page scanning for all vmcores